### DISCUSSION TITLE
Mishards throw "index out of range" error after continually search/insert for a period of time

### DISCUSSION THREAD

**Discussion by boanz** [milvus-io/milvus#5191] [Q&A and General discussion]
问题：
在执行多次（非固定次数）查询请求后，后续的查询请求会一直出现index out of range问题。
一开始是偶尔会出现几次index out of range问题，还是可以正常执行查询；跑了一段时间后，所有查询请求都会出现
index out of range问题。

mishards日志：
<img width="1123" alt="截屏2021-05-12 下午7 09 46" src="https://user-images.githubusercontent.com/34409364/117965726-ad899880-b355-11eb-834e-ee64292870d2.png">

写节点日志：
![image](https://user-images.githubusercontent.com/34409364/117971178-1d028680-b35c-11eb-9570-03d077f3ce0d.png)

读节点日志：
<img width="1312" alt="截屏2021-05-12 下午7 51 48" src="https://user-images.githubusercontent.com/34409364/117970678-81711600-b35b-11eb-99b7-c421c9c81cad.png">

修改mishards的代码，打印了从不同节点返回的查询结果，写节点返回结果的数量不等于nq * top_k（比如：192 * 128），导致出现上述问题；无论请求的nq数量为多少，返回结果数量都等于nq * 25，详见下图。
![image](https://user-images.githubusercontent.com/34409364/117966768-e413e300-b356-11eb-9bee-aa65744e072c.png)

![image](https://user-images.githubusercontent.com/34409364/117967797-107c2f00-b358-11eb-8657-647e270472aa.png)

环境：
在 Kubernetes 中部署 的Mishards 集群，包括：1个分片中间件、1 个读节点、1 个写节点。

**Comment by shengjun1985** [comment]
Excuse me, which Milvus version do you use?

**Reply by boanz** [reply]
v1.1.0

**Reply by shengjun1985** [reply]
Can you provide more information for us to reproduce the issue, such as segment_file_size, row_count, dim, metric_type, index_type?
And is it convenient for you to show us your client code?

**Reply by boanz** [reply]
Here is the code to reproduce this issue. It may take times to reproduce this issue. Maybe you can hang up the program. 
[demo.zip](https://github.com/milvus-io/milvus/files/6472315/demo.zip)

**Reply by boanz** [reply]
In order to clearly see the results returned by milvus nodes, I added this line of logging code to the mishards source code.  It may help you to find out the issue.
<img width="1107" alt="截屏2021-05-13 下午8 15 48" src="https://user-images.githubusercontent.com/34409364/118124452-2d803300-b428-11eb-9f2b-600c83bdfae3.png">

**Reply by boanz** [reply]
@shengjun1985 Sorry to border you. Can you reproduce the issue?

**Reply by shengjun1985** [reply]
Sorry for the late reply. We will investigate it soon.

**Comment by yhmo** [comment]
@boanz 
Yes, with your script, we can reproduce this error. We will investigate it.

**Reply by boanz** [reply]
Thanks.

**Comment by yhmo** [comment]
This is a bug of mishard, in the _reduce() function of /milvus/shards/mishards/service_handler.py
The 4 files code at line no.32 ~ no.36 could throw "index out of range" when there is a small segments whose row count is less than topk:
```
        if sort_f(source_diss[k - 1], diss[0]):
            return source_ids, source_diss
        if sort_f(diss[k - 1], source_diss[0]):
            return ids, diss
```
Replace the 4 lines code by the following code can avoid this error:
 ```
   src_last = len(source_diss) - 1
    if sort_f(source_diss[src_last], diss[0]):
        return source_ids, source_diss
    last = len(diss) - 1
    if sort_f(diss[last], source_diss[0]):
        return ids, diss
```

But I think this function need more code change to avoid potential bugs in certain condition.
@boanz If you know how to build mishards, you can replace the code in your local, and run "make build" and then restart the mishards cluster by "make deploy_m".

**Reply by boanz** [reply]
Thanks for your solution. 
When the issue is resolved, the code provided above will report anther issue.
![image](https://user-images.githubusercontent.com/34409364/118912949-26db4980-b95b-11eb-8a2a-6e8899f04ada.png)
It seems that the length of results returned from the mishards is not equals to the target value (nq * top_k).
If the length of results is less than the target value, will it be filled automatically?

**Reply by yhmo** [reply]
The _reduce() has a parameter "k", I think the _merge() pass a wrong parameter to "k".
Currently the "batch_len" is passed to "k":
```
 self._reduce(merge_id_results[row_index], id_batch,
                                     merge_dis_results[row_index], dis_batch,
                                     batch_len,
                                     reverse)
```

After I pass "topk" to "k", the script run long time without problem:

```
 self._reduce(merge_id_results[row_index], id_batch,
                                     merge_dis_results[row_index], dis_batch,
                                     topk,
                                     reverse) 
```

**Reply by boanz** [reply]
I've changed the code. After many executions, the issue will still come up.

**Reply by yhmo** [reply]
You mean the "index out of range" error or "Search result is not aligned" error?

**Reply by yhmo** [reply]
With the code fix #5314, I run the script for whole night, but it didn't throw error.

**Reply by boanz** [reply]
The code do fix the issue "index out of range", but the issue "Search result is not aligned" remains. 
I update the input data. It may help you to reproduce the issue.
[demo.zip](https://github.com/milvus-io/milvus/files/6521104/demo.zip)

**Reply by yhmo** [reply]
Ok, I can reproduce it.

**Reply by sworddish** [reply]
@boanz  we already noticed this problem and fixed in our own forked version, replace batch_len by topk can help

**Reply by boanz** [reply]
> @boanz we already noticed this problem and fixed in our own forked version, replace batch_len by topk can help

Thanks for your reply. The issue "Search result is not aligned"  remains even though replace batch_len by topk.

**Comment by yhmo** [comment]
After investigation, I make a code change in the same function **_reduce()**:
The original code:
```
        src_last = len(source_diss) - 1
        if sort_f(source_diss[src_last], diss[0]):
            return source_ids, source_diss
        last = len(diss) - 1
        if sort_f(diss[last], source_diss[0]):
            return ids, diss
```

My  code change:
```
        if len(ids) == 0:
            return source_ids, source_diss

        if len(source_ids) == 0:
            return ids, diss

        src_last = len(source_diss) - 1
        if sort_f(source_diss[src_last], diss[0]) and len(source_ids) >= k:
            return source_ids, source_diss
        last = len(diss) - 1
        if sort_f(diss[last], source_diss[0]) and len(ids) >= k:
            return ids, diss
```


After the code change, run the script for long time, so far I didn't encounter the error again.
@boanz Please have a try.

**Reply by boanz** [reply]
Thanks. It works.

**Comment by yhmo** [comment]
The pr #5459 has been merged into branch 1.1 for this issue.
